common.mk

```diff
@@ -308,6 +315,9 @@ append-br2-vars = $(foreach var,$(filter BR2_%,$(.VARIABLES)),$(call append-var,
 
 .PHONY: buildroot
 buildroot: optee-os
+ifeq ($(OPTEE_RUST_ENABLE),y)
+    @(cd $(OPTEE_RUST_PATH) && source setup.sh)
```

Comment:
I believe that the shell is terminated after each line, so this configuration will be rather short lived.

Reply:
Thank you for your comment. That's right, so we put all the environment variables in `br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk` and only some setup and initialization operations in `setup.sh`.

Reply:
Why is this script sourced instead of executed? The end result may be the same in this case, but it looks like an unintentional bug.

Reply:
And there is a missing TAB. Have you tested this?
```
# make buildroot OPTEE_RUST_ENABLE=y
common.mk:319: *** missing separator.  Stop.
```

Reply:
> Why is this script sourced instead of executed? The end result may be the same in this case, but it looks like an unintentional bug.

Yes, execution is better. We found `source $(EDK2_PATH)/edksetup.sh` on line 414 in common.mk, and want to maintain a unified style. Now we change `source` to `./`.

Reply:
> And there is a missing TAB. Have you tested this?
> 
> ```
> # make buildroot OPTEE_RUST_ENABLE=y
> common.mk:319: *** missing separator.  Stop.
> ```

Thanks for your correction. Sorry for my mistake. We have tested this in another directory, because my editor is not configured properly, an error occurred when pasting and copying.

Reply:
There's a subtle difference with that line. All the lines for `edk2-clean-common` target runs in the same shell.
Do you see it?

Reply:
I get it, thanks.

Reply:
How about this:
```diff
diff --git a/common.mk b/common.mk
index f72ea86..6d2d6d0 100644
--- a/common.mk
+++ b/common.mk
@@ -34,6 +34,7 @@ OPTEE_OS_PATH			?= $(ROOT)/optee_os
 OPTEE_CLIENT_PATH		?= $(ROOT)/optee_client
 OPTEE_TEST_PATH			?= $(ROOT)/optee_test
 OPTEE_EXAMPLES_PATH		?= $(ROOT)/optee_examples
+OPTEE_RUST_PATH			?= $(ROOT)/optee_rust
 BUILDROOT_TARGET_ROOT		?= $(ROOT)/out-br/target
 
 # default high verbosity. slow uarts shall specify lower if prefered
@@ -241,6 +242,13 @@ cc-option = $(strip $(call _cc-option,$(1),$(2),$(3)))
 .PHONY: all
 all:
 
+OPTEE_RUST_SET_ENV := source ~/.cargo/env
+
+optee-rust:
+ifeq ($(OPTEE_RUST_ENABLE),y)
+	cd $(OPTEE_RUST_PATH) && ./setup.sh
+endif
+
 ################################################################################
 # Build root
 ################################################################################
@@ -275,6 +283,12 @@ BR2_PACKAGE_OPTEE_EXAMPLES_EXT ?= y
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_CROSS_COMPILE ?= $(CROSS_COMPILE_S_USER)
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_SDK ?= $(OPTEE_OS_TA_DEV_KIT_DIR)
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_SITE ?= $(OPTEE_EXAMPLES_PATH)
+OPTEE_RUST_ENABLE ?= n
+ifeq ($(OPTEE_RUST_ENABLE),y)
+BR2_PACKAGE_OPTEE_RUST_EXAMPLES_EXT ?= y
+BR2_PACKAGE_OPTEE_RUST_EXAMPLES_EXT_CROSS_COMPILE ?= $(CROSS_COMPILE_S_USER)
+BR2_PACKAGE_OPTEE_RUST_EXAMPLES_EXT_SITE ?= $(OPTEE_RUST_PATH)
+endif
 # The OPTEE_OS package builds nothing, it just installs files into the
 # root FS when applicable (for example: shared libraries)
 BR2_PACKAGE_OPTEE_OS_EXT ?= y
@@ -307,7 +321,7 @@ append-var = $(call append-var_,$(1),$(2),$(if $(call y-or-n,$($(1))),,$(double-
 append-br2-vars = $(foreach var,$(filter BR2_%,$(.VARIABLES)),$(call append-var,$(var),$(1)))
 
 .PHONY: buildroot
-buildroot: optee-os
+buildroot: optee-rust optee-os
 	@mkdir -p ../out-br
 	@rm -f ../out-br/build/optee_*/.stamp_*
 	@rm -f ../out-br/extra.conf
@@ -325,7 +339,7 @@ buildroot: optee-os
 		$(DEFCONFIG_FTPM) \
 		--br-defconfig out-br/extra.conf \
 		--make-cmd $(MAKE))
-	@$(MAKE) -C ../out-br all
+	@$(OPTEE_RUST_SET_ENV) $(MAKE) -C ../out-br all
 
 .PHONY: buildroot-clean
 buildroot-clean:
```

Reply:
Hi @jforissier 

It seems that `$(OPTEE_RUST_SET_ENV)` will always execute whether `OPTEE_RUST_ENABLE` is set to `y` or not. Doing `./setup.sh` will install Rust and Cargo, so if `OPTEE_RUST_ENABLE=n` the Cargo will not be installed, then  `$HOME/.cargo/env` doesn't exist.

Reply:
@DemesneGH duh, my mistake, the definition needs to be inside the conditional of course.

Reply:
@jforissier Updated. Thanks!